### PR TITLE
[sw/lib] Delete common.h

### PR DESCRIPTION
**Pull request by mcy**
This PR deletes `common.h` after I started deleting definitions from it about a year ago. To get this over the finish line, I:
- Moved `ARRAYSIZE`, the only definition that was still useful and didn't have a libbase replacement, into `memory.h`.
- Deleted `REG16` and `REG8`, which were unused.
- Copied a definition of `REG32` into the four files that still used it (these files are proto-DIFs pending deprecation).

### CODE REVIEW

**Review by gkelly — COMMENTED** [review]
Why replicate the REG32/SETBIT/CLRBIT everywhere? I think I prefer "this bad header that's hanging around" to "we put the bad header contents into a bunch of files and now have to clean them up individually".

I'm in favor of removing `common.h` but I think we need to do that by getting rid of the things that make it bad, not by replicating them.

**Comment by mcy** [comment]
Garret: well, my thought was that those files will be deleted very soon, since DIFs exist for them at this point. I felt it was better to get rid of the header, since it's pulled in by tons of stuff that really shouldn't.

**Review by gkelly — APPROVED** [review]
Alright, that's fair. Logs look like `pinmux.c` needs the defs as well.

**Comment by sriyerg** [comment]
What is a viable solution for a test writer developing a test for a block that does not have DIFs yet?

**Comment by mcy** [comment]
@sriyerg this change should not affect that. `mmio.h` is the appropriate way to perform register accesses (any new uses of `REG32` outside of pre-existing files would have been asked to be re-written into `mmio.h` calls).

File: sw/device/lib/base/memory.h [hunk 1]
```diff
@@ -22,6 +22,22 @@
 extern "C" {
 #endif  // __cplusplus
 
+/**
+ * Computes the number of elements in the given array.
+ *
+ * Note that this can only determine the length of *static* arrays. Due to
```
Comment (alphan) [inline sw/device/lib/base/memory.h:28]: nit: ~static~ fixed-size (or fixed-length)

**Review by alphan — APPROVED** [review]
Thanks @mcy!

**Comment by sriyerg** [comment]
> @sriyerg this change should not affect that. `mmio.h` is the appropriate way to perform register accesses (any new uses of `REG32` outside of pre-existing files would have been asked to be re-written into `mmio.h` calls).

I see - you didnt replace existing REG32 calls with mmio because these sources will be removed anyway. This LGTM.

**Review by sriyerg — APPROVED** [review]

**Pull request merged**